source/extensions/filters/http/common/aws/credentials_provider_impl.cc

```diff
@@ -0,0 +1,220 @@
+    ENVOY_LOG(error, "Could not parse AWS credentials document: {}", e.what());
+    return;
+  }
+
+  const auto access_key_id = document_json->getString(ACCESS_KEY_ID, "");
+  const auto secret_access_key = document_json->getString(SECRET_ACCESS_KEY, "");
+  const auto session_token = document_json->getString(TOKEN, "");
+
+  ENVOY_LOG(debug, "Found following AWS credentials in the instance metadata: {}={}, {}={}, {}={}",
+            AWS_ACCESS_KEY_ID, access_key_id, AWS_SECRET_ACCESS_KEY,
+            secret_access_key.empty() ? "" : "*****", AWS_SESSION_TOKEN,
+            session_token.empty() ? "" : "*****");
+
+  cached_credentials_ = Credentials(access_key_id, secret_access_key, session_token);
+  last_updated_ = api_.timeSource().systemTime();
+}
+
+bool TaskRoleCredentialsProvider::needsRefresh() {
+  const auto now = api_.timeSource().systemTime();
+  return (now - last_updated_ > REFRESH_INTERVAL) ||
+         (expiration_time_ - now < REFRESH_GRACE_PERIOD);
+}
+
+void TaskRoleCredentialsProvider::refresh() {
+  ENVOY_LOG(debug, "Getting AWS credentials from the task role at URI: {}", credential_uri_);
+
+  absl::string_view host;
+  absl::string_view path;
+  Http::Utility::extractHostPathFromUri(credential_uri_, host, path);
+  const auto credential_document =
+      metadata_fetcher_(std::string(host.data(), host.size()),
+                        std::string(path.data(), path.size()), authorization_token_);
+  if (!credential_document) {
+    ENVOY_LOG(error, "Could not load AWS credentials document from the task role");
+    return;
+  }
+
+  Json::ObjectSharedPtr document_json;
+  try {
+    document_json = Json::Factory::loadFromString(credential_document.value());
+  } catch (EnvoyException& e) {
+    ENVOY_LOG(error, "Could not parse AWS credentials document from the task role: {}", e.what());
+    return;
+  }
+
+  const auto access_key_id = document_json->getString(ACCESS_KEY_ID, "");
+  const auto secret_access_key = document_json->getString(SECRET_ACCESS_KEY, "");
+  const auto session_token = document_json->getString(TOKEN, "");
+
+  ENVOY_LOG(debug, "Found following AWS credentials in the task role: {}={}, {}={}, {}={}",
+            AWS_ACCESS_KEY_ID, access_key_id, AWS_SECRET_ACCESS_KEY,
+            secret_access_key.empty() ? "" : "*****", AWS_SESSION_TOKEN,
+            session_token.empty() ? "" : "*****");
+
+  const auto expiration_str = document_json->getString(EXPIRATION, "");
+  if (!expiration_str.empty()) {
+    absl::Time expiration_time;
+    if (absl::ParseTime(EXPIRATION_FORMAT, expiration_str, &expiration_time, nullptr)) {
+      ENVOY_LOG(debug, "Task role AWS credentials expiration time: {}", expiration_str);
+      expiration_time_ = absl::ToChronoTime(expiration_time);
+    }
+  }
+
+  last_updated_ = api_.timeSource().systemTime();
+  cached_credentials_ = Credentials(access_key_id, secret_access_key, session_token);
+}
+
+Credentials CredentialsProviderChain::getCredentials() {
```

Comment:
Can we have coverage for this function? https://239673-65214191-gh.circle-artifacts.com/0/coverage/coverage.source_extensions_filters_http_common_aws_credentials_provider_impl.cc.html

Reply:
Thanks to catching that. Added tests.